### PR TITLE
[Feat] Added decompose 3x4projection matrix

### PR DESCRIPTION
**Pull request by ChetanPatil28**
### Description
This PR adds the functionality of decomposing a 3x4-projection matrix into the Camera matrix, Rotation matrix, and a Translation vector.
Since the decomposition is not unique, the camera matrix is forced to have all its elements as positive.
Related to https://github.com/kornia/kornia/issues/593

### Status
**Ready**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [x] Did you discuss the functionality or any breaking changes before ?
- [x] **Pass all tests**: did you test in local ? `make test`
- [x] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [x] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

File: kornia/geometry/epipolar/projection.py [hunk 1]
```diff
@@ -107,6 +107,44 @@ def projection_from_KRt(K: torch.Tensor, R: torch.Tensor, t: torch.Tensor) -> to
...
+    This function decomposes the Projection matrix into Camera-Matrix, Rotation Matrix and Translation vector.
+
+
+    Args:
+        P (torch.Tensor): the projection matrix with shape :math:`(B, 3, 4)`.
+
+    Returns:
+        torch.Tensor: The Camera matrix with shape :math:`(B, 3, 3)`.
+        torch.Tensor: The Rotation matrix with shape :math:`(B, 3, 3)`.
+        torch.Tensor: the Translation vector with shape :math:`(B, 3)`.
```
Comment (edgarriba) [inline kornia/geometry/epipolar/projection.py:122]: this docs won't render well since you return `Tuple`. Check this: https://github.com/kornia/kornia/blob/master/kornia/feature/matching.py#L17

File: kornia/geometry/epipolar/projection.py [hunk 2]
```diff
@@ -107,6 +107,44 @@ def projection_from_KRt(K: torch.Tensor, R: torch.Tensor, t: torch.Tensor) -> to
...
+
+    """
+
+    assert P.shape[-2:] == (3, 4), "P must be of shape [B, 3, 4]"
```
Comment (edgarriba) [inline kornia/geometry/epipolar/projection.py:126]: if we want to accept only Bx3x4 I would add an extra assertion to check `len(P.shape) == 3`.

However, could you evaluate if we could easily support `(*, 3, 4)` ? 
[`projection_from_KRt`](https://kornia.readthedocs.io/en/latest/geometry.epipolar.html#kornia.geometry.epipolar.projection_from_KRt) also accepts `Bx3x4` but now looking at the code an I believe that could be easily adapted for `(*,3,4)`. What do you think about it ?

File: kornia/geometry/epipolar/projection.py [hunk 3]
```diff
@@ -107,6 +107,44 @@ def projection_from_KRt(K: torch.Tensor, R: torch.Tensor, t: torch.Tensor) -> to
...
+
+    submat_3x3 = P[:, 0:3, 0:3]
```
Comment (edgarriba) [inline kornia/geometry/epipolar/projection.py:128]: in case we go for `(*,3,4)` -> `submat_3x3 = P[..., :3, :3]`

File: test/geometry/epipolar/test_projection.py [hunk 4]
```diff
@@ -138,3 +138,54 @@ def test_gradcheck(self, device):
...
+        assert t.shape() == (B, 3, 1)
+
+    def test_simple(self, device, dtype):
+        P = torch.Tensor([[
+            [232., 71., 267., 352.]
+            [280., 150., 140., 198.]
+            [158., 468., 237., 371.]
+        ]], device=device, dtype=dtype)
+
+        K_expected = torch.Tensor([[
```
Comment (edgarriba) [inline test/geometry/epipolar/test_projection.py:167]: torch.tensor([[

File: test/geometry/epipolar/test_projection.py [hunk 5]
```diff
@@ -138,3 +138,54 @@ def test_gradcheck(self, device):
...
+        ]], device=device, dtype=dtype)
+
+        t_expected = torch.tensor([
+            [[0.0167], [-0.1426], [-1.2950]],
+        ], device=device, dtype=dtype)
+
+        K_estimated, R_estimated, t_estimated = epi.KRt_from_projection(P)
+        assert_allclose(K_estimated, K_expected)
+        assert_allclose(R_estimated, R_expected)
+        assert_allclose(t_estimated, t_expected)
```
Comment (edgarriba) [inline test/geometry/epipolar/test_projection.py:186]: we could add an extra test to check for going forward and backward `KRt_from_projection <-> projection_from_KRt`

File: kornia/geometry/epipolar/projection.py [hunk 6]
```diff
@@ -107,6 +107,44 @@ def projection_from_KRt(K: torch.Tensor, R: torch.Tensor, t: torch.Tensor) -> to
 
     return K @ Rt
 
+def KRt_from_projection(P: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    r"""Get the intrinsics, rotation-matrix and the camera-center.
+
+    This function decomposes the Projection matrix into Camera-Matrix, Rotation Matrix and Translation vector.
+
```
Comment (edgarriba) [inline kornia/geometry/epipolar/projection.py:114]: single line

File: kornia/geometry/epipolar/projection.py [hunk 7]
```diff
@@ -107,6 +107,44 @@ def projection_from_KRt(K: torch.Tensor, R: torch.Tensor, t: torch.Tensor) -> to
...
+    """
+
+    assert P.shape[-2:] == (3, 4), "P must be of shape [B, 3, 4]"
+
+    submat_3x3 = P[:, 0:3, 0:3]
+    last_column = P[:, 0:3, 3].unsqueeze(-1)
+    t = torch.matmul(-torch.inverse(submat_3x3), last_column).squeeze(-1)
+
+    # Trick to turn QR-decomposition into RQ-decomposition
+    reverse = torch.Tensor([[0, 0, 1], [0, 1, 0], [1, 0, 0]]).unsqueeze(0)
```
Comment (edgarriba) [inline kornia/geometry/epipolar/projection.py:133]: you need to pass `device` and `dtype` otherwise some tests won't pass
`reverse = torch.tensor([[[0, 0, 1], [0, 1, 0], [1, 0, 0]]], device=P.device, dtype=P.dtype)`

File: kornia/geometry/epipolar/projection.py [hunk 8]
```diff
@@ -107,6 +107,44 @@ def projection_from_KRt(K: torch.Tensor, R: torch.Tensor, t: torch.Tensor) -> to
...
+    upper_mat = torch.matmul(reverse, torch.matmul(upper_mat.permute(0, 2, 1), reverse))
+
+    # Turning the `upper_mat's` diagonal elements to positive.
+    diagonals = torch.diagonal(upper_mat, dim1=-2, dim2=-1)
+    signs = diagonals / diagonals.abs()
+    signs_mat = torch.diag_embed(signs)
+
+    K = torch.matmul(upper_mat, signs_mat)
+    R = torch.matmul(signs_mat, ortho_mat)
+    return K, R, t
```
Comment (edgarriba) [inline kornia/geometry/epipolar/projection.py:146]: what formulation are you following for this method ? I would refer in the code.
Besides, I'd suggest to follow opencv sfm since this module is inspired on that. See: https://github.com/opencv/opencv_contrib/blob/ab2e83248fa1783d7c2055ef280d403ffcf968c8/modules/sfm/src/libmv_light/libmv/multiview/projection.cc#L32
Open to others alternatives though.

File: test/geometry/epipolar/test_projection.py [hunk 9]
```diff
@@ -138,3 +138,54 @@ def test_gradcheck(self, device):
...
+        B: int = batch_size
+        P = torch.rand(B, 3, 3, device=device, dtype=dtype)
+        K, R, t = epi.KRt_from_projection(P)
+        assert K.shape() == (B, 3, 3)
+        assert R.shape() == (B, 3, 3)
+        assert t.shape() == (B, 3, 1)
+
+    def test_simple(self, device, dtype):
+        P = torch.Tensor([[
+            [232., 71., 267., 352.]
```
Comment (edgarriba) [inline test/geometry/epipolar/test_projection.py:162]: from were did you get this numbers ?

**Review by edgarriba — CHANGES_REQUESTED** [review]

File: kornia/geometry/epipolar/projection.py (hunk 2 above)
Comment (ChetanPatil28) [reply kornia/geometry/epipolar/projection.py:126]: I'll definitely verify if my code `KRt_from_projection` can be extended to support `(*, 3, 4)` with minimal changes.
And yes, I too believe the `projection_from_KRt` can be adapted for `(*, 3, 3)`.
Anyways, I shall verify both.
Thanks for the suggestion.

File: test/geometry/epipolar/test_projection.py (hunk 9 above)
Comment (ChetanPatil28) [reply test/geometry/epipolar/test_projection.py:162]: These were the numbers generated randomly and fed into  OpenCV's function [decomposeProjectionMatrix](https://github.com/opencv/opencv/blob/60354e39018ac82d230d0d021cdc77343fc84ee8/modules/calib3d/src/calibration.cpp#L3271).
The `K`, `R` and `t` returned from this function are used to verify with my implementation.

File: kornia/geometry/epipolar/projection.py (hunk 8 above)
Comment (ChetanPatil28) [reply kornia/geometry/epipolar/projection.py:146]: I have referred to the following links for my implementation.
1. [Turning a QR decomposition to RQ decomposition](https://math.stackexchange.com/questions/1640695/rq-decomposition/1640762).

2. [Forcing all elements of Camera matrix to be positive](http://ksimek.github.io/2012/08/14/decompose/).

Currently, my implementation forces all the elements of `K` to be positive.
But, in-case of OpenCV's [decomposeProjectionMatrix](https://github.com/opencv/opencv/blob/60354e39018ac82d230d0d021cdc77343fc84ee8/modules/calib3d/src/calibration.cpp#L3271) implementation, sometimes the `last column of K` and `last row of R` turns out to be  `-ve`  wrt my code.

Have a look at the picture below.
diff-opencv-torch

This is due to the fact that inverting a specific column and corresponding row leads to the same solution.

I can change my current implementation to replicate the results as in OpenCV.

File: kornia/geometry/epipolar/projection.py (hunk 2 above)
Comment (edgarriba) [reply kornia/geometry/epipolar/projection.py:125]: sure, let us know whether you want to include this feature in this PR or after merging this one

File: kornia/geometry/epipolar/projection.py (hunk 8 above)
Comment (edgarriba) [reply kornia/geometry/epipolar/projection.py:145]: I'd prefer to replicate opencv results for convenience.

File: test/geometry/epipolar/test_projection.py (hunk 5 above)
Comment (edgarriba) [reply test/geometry/epipolar/test_projection.py:186]: add the test I mention here so that we can evaluate both directions

File: kornia/geometry/epipolar/projection.py [hunk 10]
```diff
@@ -108,6 +108,43 @@ def projection_from_KRt(K: torch.Tensor, R: torch.Tensor, t: torch.Tensor) -> to
...
+    """
+
+    assert P.shape[-2:] == (3, 4), "P must be of shape [B, 3, 4]"
+
+    submat_3x3 = P[:, 0:3, 0:3]
+    last_column = P[:, 0:3, 3].unsqueeze(-1)
+    t = torch.matmul(-torch.inverse(submat_3x3), last_column).squeeze(-1)
+
+    # Trick to turn QR-decomposition into RQ-decomposition
+    reverse = torch.Tensor([[0, 0, 1], [0, 1, 0], [1, 0, 0]], device=P.device, dtype=P.dtype).unsqueeze(0)
```
Comment (edgarriba) [inline kornia/geometry/epipolar/projection.py:132]: image

looking at the above picture from the links you shared, I think we could de compose this function other small utilities that could be very useful. E.g `flipud` which is already a numpy function and very easy to implement. See: https://numpy.org/doc/stable/reference/generated/numpy.flipud.html
```python
A = np.random.randn(2,3,5)
np.all(np.flipud(A) == A[::-1,...])
True
```

second, why not having as separate function also `kornia.rq` which implements `K.flipud` and `torch.qr`. This is a suggestion that can be included in a separated pull request

File: kornia/geometry/epipolar/projection.py [hunk 11]
```diff
@@ -108,6 +108,43 @@ def projection_from_KRt(K: torch.Tensor, R: torch.Tensor, t: torch.Tensor) -> to
...
+    submat_3x3 = torch.matmul(reverse, submat_3x3).permute(0, 2, 1)
+    ortho_mat, upper_mat = torch.qr(submat_3x3)
+    ortho_mat = torch.matmul(reverse, ortho_mat.permute(0, 2, 1))
+    upper_mat = torch.matmul(reverse, torch.matmul(upper_mat.permute(0, 2, 1), reverse))
+
+    # Turning the `upper_mat's` diagonal elements to positive.
+    diagonals = torch.diagonal(upper_mat, dim1=-2, dim2=-1)
+    signs = diagonals / diagonals.abs()
```
Comment (edgarriba) [inline kornia/geometry/epipolar/projection.py:140]: you might need to add an epsilon value to the denominator. Besides, pytorch already provides [`torch.sign`](https://pytorch.org/docs/stable/generated/torch.sign.html?highlight=sign#torch.sign), but not quite sure about its differentiability property

File: kornia/geometry/epipolar/projection.py (hunk 2 above)
Comment (ChetanPatil28) [reply kornia/geometry/epipolar/projection.py:125]: I was thinking of implementing this extension from `B` to `*` after emerging this one.
Currently, I'll add the assertion for `len(P.shape) == 3`

File: kornia/geometry/epipolar/projection.py (hunk 11 above)
Comment (ChetanPatil28) [reply kornia/geometry/epipolar/projection.py:140]: Yes, `torch.sign` is differentiable. I have replaced the previous code with `torch.sign()`. 
Also added the `epsilon` to `diagonals` to prevent elements of `signs` being `zero.`

File: kornia/geometry/epipolar/projection.py (hunk 8 above)
Comment (ChetanPatil28) [reply kornia/geometry/epipolar/projection.py:145]: Sure, I'll update the code.

Comment (ChetanPatil28) [reply kornia/geometry/epipolar/projection.py:150]: I will be implementing the algorithm as written in https://github.com/opencv/opencv_contrib/blob/ab2e83248fa1783d7c2055ef280d403ffcf968c8/modules/sfm/src/libmv_light/libmv/multiview/projection.cc#L32

Some of the test-cases did not pass in the previous algorithm, hence I had to write code as per the link pasted above.

File: kornia/geometry/epipolar/projection.py [hunk 12]
```diff
@@ -123,6 +123,9 @@ def KRt_from_projection(P: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, to
     """
 
     assert P.shape[-2:] == (3, 4), "P must be of shape [B, 3, 4]"
+    assert len(P.shape) == 3
+
+    epsilon = torch.tensor([0.00001], device=P.device, dtype=P.dtype)
```
Comment (edgarriba) [inline kornia/geometry/epipolar/projection.py:128]: I would suggest to define epsilon in the function signature as `float`. Something like this:

`def KRt_from_projection(P: torch.Tensor, eps: float = 1e-6) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:`

File: test/geometry/epipolar/test_projection.py (hunk 5 above)
Comment (ChetanPatil28) [reply test/geometry/epipolar/test_projection.py:186]: Hello @edgarriba , since I'm going with replication  of result wrt [this implementation](https://github.com/opencv/opencv_contrib/blob/ab2e83248fa1783d7c2055ef280d403ffcf968c8/modules/sfm/src/libmv_light/libmv/multiview/projection.cc#L32),  the `C-code` actually normalizes the `K-matrix` in the end (to make the last element to be 1). 
And if this `K` is fed to `projection_from_KRt()` as a sanity-check, we will get different result for `P`

So, I will have to leave the `K-matrix` un-normalized to cope with getting the exact `P` out of `projection_from_KRt()`

File: kornia/geometry/epipolar/projection.py (hunk 10 above)
Comment (ChetanPatil28) [reply kornia/geometry/epipolar/projection.py:135]: Yes, I will break the whole function into smaller utilities, but in the next PR.

**Review by edgarriba — APPROVED** [review]
